[SDPA] Call _sdp_attention  in nn.functional.mha

drisspg opened the pull request:
# Summary
Replaces the the inline block of code in nn.funcitonal.mha with `_scaled_dot_product_attention`. This function allows the fused kernels to be called if all the required input conditions are met.

cc @VitalyFedyunin @ngimel

drisspg commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -741,10 +740,12 @@ std::tuple<Tensor, Tensor> _scaled_dot_product_attention_math(
     // Scale q,k before matmul for stability see https://tinyurl.com/sudb9s96 for math
-    const double scaling_factor = ::sqrt(::sqrt(static_cast<double>(embed_size)));
+    const auto embed_size = SymFloat(query_.sym_size(-1));
+    // const double scaling_factor = ::sqrt(::sqrt(static_cast<double>(embed_size)));
```
TODO Different options I was working through, need to remove before landing

drisspg commented on aten/src/ATen/native/transformers/cuda/sdp_utils.h:
```diff
@@ -162,6 +164,25 @@ inline bool check_head_dim_size(sdp_params params, bool debug) {
+  const int64_t query_size_last = params.query.size(-1);
+  if (!(query_size_last == params.key.size(-1) &&
+        query_size_last == params.value.size(-1) && query_size_last >= 8)) {
+    TORCH_CHECK(
```
For head_dim size less than 8 efficient_attention will throw with unaligned memory format

drisspg commented on c10/core/SymFloat.cpp:
```diff
@@ -70,6 +71,15 @@ std::ostream& operator<<(std::ostream& os, const SymFloat& s) {
   return os;
 }
 
+SymFloat SymFloat::sqrt() const {
```
Needs double checking and placement guidance

drisspg commented on torch/nn/functional.py:
```diff
@@ -5175,19 +5175,20 @@ def multi_head_attention_forward(
-    attn_output_weights = softmax(attn_output_weights, dim=-1)
-    if dropout_p > 0.0:
-        attn_output_weights = dropout(attn_output_weights, p=dropout_p)
+        if attn_mask.size(0) == 1:
```
hacky.. the API between SDP vs the nn.funcitonal.mha has a decent amount of impedance mismatch requiring all this transposing and viewing fluff. Ideally this would also work with nested tensors out of the box but more thought needs to be takin to do so

drisspg commented:
@BowenBao @abock Sorry for pinging you directly but I am getting a test failure for:
```
test/onnx/test_models_onnxruntime.py::TestModelsONNXRuntime_is_script_False::test_transformer_encoder - 
torch.onnx.errors.UnsupportedOperatorError: Exporting the operator 'aten::_scaled_dot_product_attention' to ONNX opset version 14 is not supported. 
Please feel free to request support or submit a pull request on PyTorch GitHub: https://github.com/pytorch/pytorch/issues.
```
Which can be found on the hud: https://hud.pytorch.org/pr/89470

I am not sure what are the next steps to enable ONNX support. I tried reading through the wiki but didn't find anything very fruitful. Any guidance would be much appreciated

bdhirsh commented on aten/src/ATen/native/transformers/cuda/sdp_utils.h:
```diff
@@ -162,6 +164,25 @@ inline bool check_head_dim_size(sdp_params params, bool debug) {
   return true;
 }
 
+inline bool check_head_dim_size_mem_efficient(sdp_params params, bool debug) {
```
does this code not need to be traced, so it doesn't need to be symintified? (e.g. it's only invoked somewhere inside a cuda kernel)

drisspg replied:
hmm thats a good question: 
This code is used to check if we can run the fused kernels:
https://github.com/pytorch/pytorch/blob/ed41a7fb681ff51129763213042da0be4fbd1a56/aten/src/ATen/native/transformers/cuda/attention.cu#LL779C29-L779C29

In particular it is uesd in the cuda implementation of the native function _fused_sdp_choice which is in turn used here:
https://github.com/pytorch/pytorch/blob/ed41a7fb681ff51129763213042da0be4fbd1a56/aten/src/ATen/native/transformers/attention.cpp#L693

bdhirsh replied:
yep - you're right, it looks like this code doesn't need to be traced (it's only called from inside of cuda kernels)

bdhirsh commented on c10/core/SymFloat.h:
```diff
@@ -40,6 +40,9 @@ class C10_API SymFloat {
   SymFloat operator/(const SymFloat&) const;
 
+  // Need guidance on where to put this code
+  SymFloat sqrt() const;
```
Have you tried writing something like:


```
namespace std {
  template <>
  SymFloat sqrt<SymFloat>(const SymFloat& self) {
    ...
  }
}
```

Not sure if this would work exactly, but maybe we can get `std::sqrt()` to Just Work on symfloats, so we don't need any code changes in the future

drisspg replied:
Can't seem to match the function template specialization  to a function template. Tried for both std:: and c10::complex

bdhirsh commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -741,10 +740,12 @@ std::tuple<Tensor, Tensor> _scaled_dot_product_attention_math(
 
     // Scale q,k before matmul for stability see https://tinyurl.com/sudb9s96 for math
-    const double scaling_factor = ::sqrt(::sqrt(static_cast<double>(embed_size)));
+    const auto embed_size = SymFloat(query_.sym_size(-1));
```
The manual cast with `SymFloat(...)` seems ok to me!

mikekgfb commented on torch/nn/functional.py:
```diff
@@ -5173,21 +5173,26 @@ def multi_head_attention_forward(
+
+    attn_output_2, attn_output_weights_2 = _scaled_dot_product_attention(
+        q, k, v, attn_mask, dropout_p, need_weights, False)
+    attn_output_2 = attn_output_2.transpose(1, 2).transpose(0, 1).contiguous().view(bsz * tgt_len, embed_dim)
```
The transpose(0 calls necessary to pull this off might cost us significant overheads.

There's an impedeance mismatch not just between functional.MHA and sdp_attention, but also between nn.MHA and functional.MHA.

nn.MHA treats batch_first as preferred format, but functional.MHA appears to only support not batch_first. I see two options:
1 - pass in the bacth_first variable (but that's somewhat non-preferred because we might end up with control flow as a function of an input, the input being batch_first.)
2 - have an inverted polarity MHA which prefers batch_first and use that one internally, then have nn.MHA and legacy compatibility functional.MHA call the new implementation - with a possible path to deprecating the old interface eventually)

drisspg commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot revert -m "breaking internal builds" -c ghfirst

jeanschmidt commented:
This PR needed to be reverted due it is breaking internal CI tests for the diff D41586942. Could not identify the cause. Reaching the author directly with more details.

jeanschmidt commented:
@pytorchbot revert -m "breaking internal builds" -c ghfirst

drisspg committed:
Call _sdp_attention in nn.functional.mha (#6038)

Summary:
X-link: https://github.com/pytorch/glow/pull/6038

Replaces the the inline block of code in nn.funcitonal.mha with `_scaled_dot_product_attention`. This function allows the fused kernels to be called if all the required input conditions are met.

cc VitalyFedyunin ngimel

Pull Request resolved: https://github.com/pytorch/pytorch/pull/89470

Reviewed By: mostafaelhoushi, cpuhrsch

Differential Revision: D41625335

Pulled By: drisspg

fbshipit-source-id: 1723c11739fc73963bd9be8dc04f45e5abda79c0
